aws/hll/aws-s3-transfer-manager/src/download/discovery.rs

```diff
@@ -0,0 +1,337 @@
+pub(super) async fn discover_obj(
+    handle: &DownloadHandle,
+    request: &DownloadRequest,
+) -> Result<ObjectDiscovery, error::TransferError> {
+    let strategy = ObjectDiscoveryStrategy::from_request(request)?;
+    match strategy {
+        ObjectDiscoveryStrategy::HeadObject(byte_range) => {
+            discover_obj_with_head(handle, request, byte_range).await
+        }
+        ObjectDiscoveryStrategy::FirstPart => {
+            let r = request.input.clone().part_number(1);
+            discover_obj_with_get(handle, r, None).await
+        }
+        ObjectDiscoveryStrategy::RangedGet(range) => {
+            let byte_range = match range.as_ref() {
+                Some(r) => ByteRange::Inclusive(
+                    *r.start(),
+                    cmp::min(*r.start() + handle.target_part_size - 1, *r.end()),
+                ),
+                None => ByteRange::Inclusive(0, handle.target_part_size - 1),
+            };
+            let r = request
+                .input
+                .clone()
+                .set_part_number(None)
+                .range(header::Range::bytes(byte_range));
+
+            discover_obj_with_get(handle, r, range).await
+        }
+    }
+}
+
+async fn discover_obj_with_head(
+    handle: &DownloadHandle,
+    request: &DownloadRequest,
+    byte_range: Option<ByteRange>,
+) -> Result<ObjectDiscovery, error::TransferError> {
+    let meta: ObjectMetadata = handle
+        .client
+        .head_object()
+        .set_bucket(request.input.get_bucket().clone())
+        .set_key(request.input.get_key().clone())
+        .send()
+        .await
+        .map_err(|e| error::DownloadError::DiscoverFailed(e.into()))?
+        .into();
+
+    let remaining = match byte_range {
+        Some(range) => match range {
+            ByteRange::Inclusive(start, end) => start..=end,
+            ByteRange::AllFrom(start) => start..=meta.total_size(),
+            ByteRange::Last(n) => (meta.total_size() - n + 1)..=meta.total_size(),
+        },
+        None => 0..=meta.total_size(),
+    };
+
+    Ok(ObjectDiscovery {
+        remaining,
+        meta,
+        initial_chunk: None,
+    })
+}
+
+async fn discover_obj_with_get(
+    handle: &DownloadHandle,
+    request: GetObjectInputBuilder,
+    range: Option<RangeInclusive<u64>>,
+) -> Result<ObjectDiscovery, error::TransferError> {
+    let resp = request.send_with(&handle.client).await;
+
+    if resp.is_err() {
+        // TODO - deal with empty file errors, see https://github.com/awslabs/aws-c-s3/blob/v0.5.7/source/s3_auto_ranged_get.c#L147-L153
+    }
+
+    let mut resp = resp.map_err(|e| error::DownloadError::DiscoverFailed(e.into()))?;
+
+    // take the body so we can convert the metadata
+    let empty_stream = ByteStream::new(SdkBody::empty());
+    let body = mem::replace(&mut resp.body, empty_stream);
+
+    let data = body
+        .collect()
+        .await
+        .map_err(|e| error::DownloadError::DiscoverFailed(e.into()))?;
```

Comment:
ideally this error's constructor would take an `impl Error` or whatever the `into` is doing. Then this line could be

```
.map_err(error::DownloadError::DiscoverFailed)?;
```